### DISCUSSION TITLE
In a store, how do you access $auth?

### DISCUSSION THREAD

**Discussion by nathanchase** [Teranode/nuxt-module-alternatives#54] [Q&A] [answered]
I can get to `$auth` by way of `useNuxtApp()` in components and composables, but how would you access it in a [Pinia](https://github.com/vuejs/pinia) store?

In components, I usually just do:
```js
const nuxtApp = useNuxtApp();
const authUser = nuxtApp.$auth.user;
```
but in a `stores/store.ts`, `useNuxtApp` is not available, so I'm not quite sure how to get to the context?
```
[nuxt] [request error] nuxt instance unavailable
  at Module.useNuxtApp (./.nuxt/dist/server/server.mjs:520:13)
  at $id_C1ltjrBOrk (./.nuxt/dist/server/server.mjs:10547:39)
  at async __instantiateModule__ (./.nuxt/dist/server/server.mjs:32909:3)
```

I notice too that this version of the auth module is auto-creating a Pinia store called `auth` with all of the data objects, but I don't know how to access the store itself - usually it would be with some sort of `const authStore = useAuthStore();`.

Reference: https://github.com/vuejs/pinia/pull/1433#issuecomment-1192495203

**Comment by Denoder** [comment]
`$auth.$storage.getStore()`

**Reply by nathanchase** [reply]
I'm unable to access `$auth` though?
```
[nuxt] [request error] $auth is not defined
  at $id_C1ltjrBOrk (./.nuxt/dist/server/server.mjs:8705:19)
  at async __instantiateModule__ (./.nuxt/dist/server/server.mjs:32973:3)
  ```
![image](https://user-images.githubusercontent.com/5963574/180460936-a91846fc-f7f5-4018-8c7a-0201e9960965.png)

**Reply by Denoder** [reply]
is that without useNuxtApp()?

**Reply by nathanchase** [reply]
That's what I'm saying - inside a Pinia store there is no `useNuxtApp()` or `this` or `$auth`.

**Reply by Denoder** [reply]
I've updated the npm package, try to access `$auth` without useNuxtApp()

**Reply by nathanchase** [reply]
Result:
```
pinia.mjs:1648 Uncaught (in promise) Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?
	const pinia = createPinia()
	app.use(pinia)
This will fail in production.
    at useStore (pinia.mjs:1648:19)
    at setup (default.vue:22:1)
    at callWithErrorHandling (runtime-core.esm-bundler.js:155:22)
    at setupStatefulComponent (runtime-core.esm-bundler.js:7165:29)
    at setupComponent (runtime-core.esm-bundler.js:7120:11)
    at mountComponent (runtime-core.esm-bundler.js:5473:13)
    at hydrateNode (runtime-core.esm-bundler.js:4667:21)
    at hydrateSubTree (runtime-core.esm-bundler.js:5559:25)
    at runtime-core.esm-bundler.js:5570:56
```

**Reply by Denoder** [reply]
can you try to see if `$auth.user` works for you?

**Reply by Denoder** [reply]
@nathanchase can you also show me what your modules array look like in your nuxt config?

**Reply by nathanchase** [reply]
```
modules: [
    '@averjs/nuxt-compression', // https://github.com/exreplay/nuxt-compression
    '@vueuse/nuxt', // https://github.com/vueuse/vueuse
    '@kevinmarrec/nuxt-pwa', // https://github.com/kevinmarrec/nuxt-pwa-module (check for official version Nuxt 3 support)
    '@nuxt/image-edge', // https://v1.image.nuxtjs.org/ (check for improved Nuxt 3 support)
    '@nuxtjs/device', // https://github.com/Redemption198/device-module (check for official version Nuxt 3 support)
    '@nuxtjs/stylelint-module', // https://github.com/nuxt-community/stylelint-module
    '@nuxtjs-alt/auth', // https://github.com/Teranode/nuxt-module-alternatives (check for official version Nuxt 3 support)
    '@nuxtjs-alt/axios', // https://github.com/Teranode/nuxt-module-alternatives (check for official version Nuxt 3 support)
    ['@pinia/nuxt', { autoImports: ['defineStore'] }], // https://github.com/vuejs/pinia
  ],
```

**Reply by Denoder** [reply]
Im gonna need a reproducible link for me to troubleshoot what the issue is cause I'm not receiving this error

**Reply by nathanchase** [reply]
Here's a bare bones reproduction: 
https://stackblitz.com/edit/github-ygh9qg-ub3dg9?file=stores%2Fstore.ts

**Reply by Denoder** [reply]
@nathanchase i'm still not recieving the issue (also pnpm didn't install so I changed it to yarn)
https://stackblitz.com/edit/github-ygh9qg-1kfzhg?file=pages/index.vue

**Reply by nathanchase** [reply]
You see that it can't find $auth though, right?
![image](https://user-images.githubusercontent.com/5963574/180504126-19c6c4dd-2d01-47ba-8746-0952525e0bdd.png)

**Reply by Denoder** [reply]
:/ but wouldn't $auth be located in `useNuxtApp()`, I'm not sure what the issue here is if it functions

![image](https://user-images.githubusercontent.com/1527149/180506490-fa8dfdd2-fb77-4505-925a-56720b52a98e.png)

**Reply by nathanchase** [reply]
Well, if I use `import { useNuxtApp } from '#imports';` then I get a new error:
```
[h3] [unhandled] H3Error: nuxt instance unavailable
    at createError (file:///home/zampa/nuxt3-jul22/flickchart/node_modules/.pnpm/h3@0.7.12/node_modules/h3/dist/index.mjs:191:15)
    at Server.nodeHandler (file:///home/zampa/nuxt3-jul22/flickchart/node_modules/.pnpm/h3@0.7.12/node_modules/h3/dist/index.mjs:381:21) {
  statusCode: 500,
  fatal: false,
  unhandled: true,
  statusMessage: 'Internal Server Error'
}
```
I think this is a matter of the nuxt instance not being accessible to the server-side and that's why there's no way to get to $auth or use useNuxtApp() in a store.

**Reply by nathanchase** [reply]
We've solved it! It's a matter of WHERE you declare `useNuxtApp()` - it CAN'T be at the top. It HAS to be inside of each individual getter or action. https://github.com/nuxt/framework/discussions/6072#discussioncomment-3207534

**Comment by nathanchase** [comment, marked as answer]
This looks to be the answer: `useNuxtApp()` must be declared INSIDE **_each_** Pinia getter or action. You CANNOT just declare it at the top of a store file. https://github.com/nuxt/framework/discussions/6072#discussioncomment-3207534
